src/LeaseManager.cc

```diff
@@ -0,0 +1,307 @@
+
+    // Populate the lease information if it is found.
+    LeaseMap::iterator leaseEntry = leaseMap.find(leaseId);
+    if (leaseEntry != leaseMap.end()) {
+        clientLease.leaseId = leaseId;
+        clientLease.leaseTerm = leaseEntry->second;
+    } else {
+        clientLease.leaseId = 0;
+        clientLease.leaseTerm = 0;
+    }
+
+    clientLease.timestamp = clock.getTime();
+
+    return clientLease;
+}
+
+/**
+ * Recover lease information from external storage.  This should only be called
+ * once after the construction of the Lease Manager.
+ */
+void
+LeaseManager::recover()
+{
+    Lock lock(mutex);
+
+    // Fetch all lease information from external storage.
+    vector<ExternalStorage::Object> objects;
+    context->externalStorage->getChildren(STORAGE_PREFIX.c_str(), &objects);
+
+    // We can safely make the approximation that every allocated lease was
+    // issued with the exception of the largest allocated leaseId.
+    foreach (ExternalStorage::Object& object, objects) {
+        try {
+            std::string name = object.name;
+            uint64_t leaseId = std::stoull(name);
+
+            // Keep track of the largest allocated leaseId and only auto "renew"
+            // the other leaseIds.
+            if (maxAllocatedLeaseId == 0) {
+                maxAllocatedLeaseId = leaseId;
+                continue;
+            } else if (leaseId > maxAllocatedLeaseId) {
+                uint64_t temp = maxAllocatedLeaseId;
+                maxAllocatedLeaseId = leaseId;
+                leaseId = temp;
+            }
+
+            uint64_t leaseTerm = clock.getTime() + LeaseCommon::LEASE_TERM_US;
+            leaseMap[leaseId] = leaseTerm;
+            expirationOrder.insert({leaseTerm, leaseId});
+        } catch (std::invalid_argument& e) {
+            LOG(ERROR, "Unable to recover lease: %s", object.name);
+        }
+    }
+
+    if (maxAllocatedLeaseId > 0) {
+        lastIssuedLeaseId = maxAllocatedLeaseId - 1;
+    }
+}
+
+/**
+ * Attempts to renew a lease with leaseId. If the requested leaseId is still
+ * live the lease is renewed and its term is extended; if not a new leaseId is
+ * issued and returned.
+ *
+ * \param leaseId
+ *      The leaseId that a client wishes to renew if possible.  Note that
+ *      leaseId = 0 is implicitly invalid and will always cause a new lease to
+ *      be returned.
+ * \return
+ *      leaseId and leaseTerm for the renewed lease or new lease.
+ */
+WireFormat::ClientLease
+LeaseManager::renewLease(uint64_t leaseId)
+{
+    Lock lock(mutex);
+    WireFormat::ClientLease clientLease = renewLeaseInternal(leaseId, lock);
+
+    // Poke the preallocator if we are getting close to running out of leases.
+    if (maxAllocatedLeaseId - lastIssuedLeaseId < PREALLOCATION_LIMIT / 4) {
+        preallocator.start(0);
+    }
+
+    return clientLease;
+}
+
+/**
+ * Start background timers to perform preallocation and cleaning and also start
+ * the cluster clock updater.
```

Comment:
Why is this a separate method, rather than just being part of the constructor? I'm wondering if there's some design decision that you need to document above.